KEYCLOAK-18445 Add support for cross-site model tests

martin-kanis opened the pull request.

martin-kanis committed:
KEYCLOAK-18445 Add support for cross-site model tests

mhajas commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java:
```diff
@@ -143,18 +147,67 @@ public void testUserSessionInitializerWithDeletingClient() {
+                    });
+                    userSessionId.set(userSessionModel.getId());
+                } else {
+                    if (siteName.get().equals(CONFIG.scope("connectionsInfinispan", "default").get("siteName", "site-1"))) {
```
This if seems to be redundant.

martin-kanis replied:
Good point. I simplified it.

mhajas commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java:
```diff
@@ -157,7 +157,6 @@ public void testPersistenceWithLoad() {
             // Assert offline sessions
             RealmModel realm = session.realms().getRealm(realmId);
             List<UserSessionModel> loadedSessions = loadPersistedSessionsPaginated(session, true, 2, 2, 3);
-            assertSessions(loadedSessions, new String[] { origSessions[0].getId(), origSessions[1].getId(), origSessions[2].getId() });
```
Should we check for size here?

martin-kanis replied:
I returned this back. Not sure why I removed it.

mhajas reviewed (commented):
Great work @martin-kanis. This will really help with HotRod testing! I added some comments/questions to the code.

martin-kanis commented on model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java:
```diff
@@ -205,7 +196,8 @@ protected void initEmbedded() {
             String jgroupsUdpMcastAddr = config.get("jgroupsUdpMcastAddr", System.getProperty(InfinispanConnectionProvider.JGROUPS_UDP_MCAST_ADDR));
-            configureTransport(gcb, topologyInfo.getMyNodeName(), topologyInfo.getMySiteName(), jgroupsUdpMcastAddr);
+            configureTransport(gcb, topologyInfo.getMyNodeName(), topologyInfo.getMySiteName(), jgroupsUdpMcastAddr,
+                    "default-configs/default-keycloak-jgroups-udp.xml", this.getClass().getClassLoader());
```
Good point. No removed it and used the classloader in `InfinispanUtil`.

martin-kanis commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/InfinispanUtil.java:
```diff
@@ -92,4 +113,142 @@ public static long toHotrodTimeMs(BasicCache ispnCache, long lifespanOrigMs) {
         return lifespanOrigMs;
     }
 
+    private static final Object CHANNEL_INIT_SYNCHRONIZER = new Object();
```
No, I just moved those methods from `DefaultInfinispanConnectionProviderFactory`.

martin-kanis commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/CrossDCInfinispan.java:
```diff
@@ -0,0 +1,91 @@
+
+    private static final String SITE_2_MCAST_ADDR = "228.6.7.8";
+
+    private Object lock = new Object();
```
Added

martin-kanis commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java:
```diff
@@ -143,18 +147,67 @@ public void testUserSessionInitializerWithDeletingClient() {
+    private AtomicReferenceArray createSessionsInPersisterOnly() {
         UserSessionModel[] origSessions = inComittedTransaction(session -> { return UserSessionPersisterProviderTest.createSessions(session, realmId); });
-        String[] res = new String[origSessions.length];
+        AtomicReferenceArray<String> res = new AtomicReferenceArray(origSessions.length);
```
Changed

martin-kanis commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java:
```diff
@@ -143,18 +147,67 @@ public void testUserSessionInitializerWithDeletingClient() {
+            }
+        });
+
+        assertThat(containsSession.get(), everyItem(is(true)));
```
Yeah I added a check for size and also little logic to distinguish runs with Hot Rod and without Hot Rod (final size depends on that).

martin-kanis commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java:
```diff
@@ -109,17 +116,16 @@ public void testUserSessionInitializer() {
     @Test
     public void testUserSessionInitializerWithDeletingClient() {
-        String[] origSessionIds = createSessionsInPersisterOnly();
+        AtomicReferenceArray<String> origSessionIds = createSessionsInPersisterOnly();
```
Yeah it's not needed. Reverted.

martin-kanis commented on model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionAdapter.java:
```diff
@@ -111,22 +111,22 @@ public boolean isOffline() {
-                        removedClientUUIDS.add(clientUUID);
-                    }
-                });
+        // to avoid concurrentModificationException
```
You can reproduce the issue like this:

- revert changes in `MapUserSessionAdapter`
- For example from `UserSessionProviderModelTest#testExpiredClientSessions` in last `inComittedTransaction` call `userSession.getAuthenticatedClientSessions().entrySet().stream().forEach(c -> {});`
- `mvn test -f testsuite/model/pom.xml -Pmap -Dtest=UserSessionProviderModelTest#testExpiredClientSessions`

```
java.util.ConcurrentModificationException
	at java.base/java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:719)
	at java.base/java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:751)
	at java.base/java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:749)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.keycloak.models.map.userSession.MapUserSessionAdapter.getAuthenticatedClientSessions(MapUserSessionAdapter.java:116)
```
I'm not sure why it's happening.

mhajas replied:
I found out the reason of the `CME`. The problem is, that `session.sessions().getClientSession()` [udpates](https://github.com/keycloak/keycloak/blob/master/model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java#L116) `entity.getAuthenticatedClientSessions()` map when the clientSession expires. 

We need to do the `getClientSession` call outside of looping through `authenticatedClientSessions`. I cannot come up with any nice solution, so the copy of `authenticatedClientSessions` is probably the best. But if you can find anything better it would be great.

martin-kanis commented on services/src/main/java/org/keycloak/executors/DefaultExecutorsProviderFactory.java:
```diff
@@ -189,7 +189,7 @@ protected ExecutorService createPoolEmbedded(String taskType, KeycloakSession se
             return new ThreadPoolExecutor(min, max,
                     60L, TimeUnit.SECONDS,
-                    new SynchronousQueue<Runnable>(),
+                    new ArrayBlockingQueue<>(1024),
```
I can't remember the exact error message but it was related to `InfinispanNotificationsManager` and some cluster listener. Maybe @hmlnarik can remember more?

Anyway, I reverted the change and I don't see the error again. Maybe it was present only in earlier phases of this PR.

martin-kanis replied:
The problem showed up again after I reverted the change. Here is the tests [output].(https://pipelines.actions.githubusercontent.com/ENJd2H9lx5qREjW4so5u3q3ePn6aZgA0m90QfCJXr9DcNaynQj/_apis/pipelines/1/runs/492/signedlogcontent/21?urlExpires=2021-10-04T10%3A17%3A43.7621894Z&urlSigningMethod=HMACV1&urlSignature=MlpIVVmvvG3vLpPdAMrKxmjptOtPsYXWmW%2B2KtyyGwA%3D)

hmlnarik replied:
> Why is this change necessary?
> 
> If I understand correctly, this means, that when 1024 tasks are waiting for execution a `RejectedExecutionException` exception is thrown. Should we catch this exception for example here: https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/protocol/saml/SamlService.java#L387?

No, there are many other places that use this functionality, and the exhausted pool should be visible to the administrator to take appropriate action in configuring the pool. FTR, the default queue in Wildfly is unbounded.

martin-kanis commented:
@mhajas Thank you for the feedback! I addressed your comments and answered your questions. 

CrossDC tests failed but I don't see how that could be related to this PR.

mhajas reviewed (dismissed):
Thank you for the fixes @martin-kanis! I am approving, but would be great to get some feedback from @hmlnarik regarding `ArrayBlockingQueue` vs `SynchronousQueue` [1] before merging.

[1] https://github.com/keycloak/keycloak/pull/8267#discussion_r709270460

hmlnarik commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java:
```diff
@@ -594,8 +594,8 @@ public static void assertSession(UserSessionModel session, UserModel user, Strin
-        assertTrue(session.getStarted() >= started - 1 && session.getStarted() <= started + 1);
-        assertTrue(session.getLastSessionRefresh() >= lastRefresh - 1 && session.getLastSessionRefresh() <= lastRefresh + 1);
+        assertTrue(session.getStarted() >= started - 2 && session.getStarted() <= started + 2);
+        assertTrue(session.getLastSessionRefresh() >= lastRefresh - 2 && session.getLastSessionRefresh() <= lastRefresh + 2);
```
Why is this change necessary?

martin-kanis replied:
It started to fail after I changed `userSessionprovider.removeLocalUserSessions()` to `reinitializeKeycloakSessionFactory` in `createSessionsInPersisterOnly()`. 
At the end I decided to fix it more nicely by capturing `started` directly from the created offline session rather than from `Time.currentTime()` after the `createSessionsInPersisterOnly() is called`.

martin-kanis commented on testsuite/model/src/main/java/org/keycloak/testsuite/model/KeycloakModelParameters.java:
```diff
@@ -67,4 +67,11 @@ public Statement instanceRule(Statement base, Description description) {
+
+    public void afterClass() {
+
+    }
```
I changed `suiteRule` to `beforeSuite` and removed `afterClass` and used `classRule`.

martin-kanis commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/InfinispanUtil.java:
```diff
@@ -37,6 +56,8 @@
  */
 public class InfinispanUtil {
```
moved

martin-kanis commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/InfinispanUtil.java:
```diff
@@ -92,4 +113,142 @@ public static long toHotrodTimeMs(BasicCache ispnCache, long lifespanOrigMs) {
+    private static final Object CHANNEL_INIT_SYNCHRONIZER = new Object();
+
+    public static void configureTransport(GlobalConfigurationBuilder gcb, String nodeName, String siteName, String jgroupsUdpMcastAddr,
+                                      String jgroupsUdpConfigPath) {
```
changed

martin-kanis commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java:
```diff
@@ -400,6 +406,13 @@ public void cleanEnvironment() {
+            kmp.afterClass();
+        }
+    }
+
```
done

martin-kanis commented on testsuite/model/src/main/resources/hotRod/hotRod1.xml:
```diff
@@ -0,0 +1,43 @@
+<infinispan>
```
renamed

martin-kanis commented on testsuite/model/src/main/java/org/keycloak/testsuite/model/HotRodServerRule.java:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+    private void createKeycloakCaches(boolean async, String ...cache) {
```
fixed

hmlnarik reviewed (dismissed):
Thank you for the PR!
